pkg/kubectl/cmd/delete/delete.go

```diff
@@ -143,12 +143,122 @@ func NewCmdDelete(f cmdutil.Factory, streams genericclioptions.IOStreams) *cobra
 	return cmd
 }
 
+/* Get cluster scoped resource type list */
+func getClsScopedRes(f cmdutil.Factory) map[string]bool {
+	resMap := make(map[string]bool)
+
+	c, err := f.ToDiscoveryClient()
+	if err != nil {
+		return resMap
+	}
+
+	lists, err := c.ServerPreferredResources()
+	if err != nil {
+		return resMap
+	}
+
+	for _, list := range lists {
+		for _, res := range list.APIResources {
+			if res.Namespaced {
+				continue
+			}
+
+			/* Just care about cluster scoped resource types */
+			if len(res.Name) > 0 {
+				resMap[res.Name] = true
+			}
+
+			if len(res.SingularName) > 0 {
+				resMap[res.SingularName] = true
+			}
+
+			/*
+			 * Kind: Pod,       Name: pods
+			 * Kind: Namespace, Name: namespaces
+			 *
+			 * The end user could specified pod or namespace in the command line
+			 */
+			if len(res.Kind) > 0 {
+				resMap[strings.ToLower(res.Kind)] = true
+			}
+
+			/* Put short names in the return list */
+			for _, s := range res.ShortNames {
+				if len(s) > 0 {
+					resMap[s] = true
+				}
+			}
+		}
+	}
+
+	return resMap
+}
+
+func checkNamespaceWithArgs(f cmdutil.Factory, args string, cmd *cobra.Command) error {
+	if len(args) == 0 {
+		return nil
+	}
+
+	nsSpecified := cmd.Flags().Changed("namespace")
+	allNsSpecified := cmd.Flags().Changed("all-namespaces")
+
+	if !nsSpecified && !allNsSpecified {
+		return nil
+	}
+
+	if nsSpecified && allNsSpecified {
+		return fmt.Errorf("--namespace (-n) and --all-namespaces cannot be specified together")
+	}
+
+	clsScopedRes := getClsScopedRes(f)
+	if len(clsScopedRes) == 0 {
+		return nil
+	}
+
+	/* Check parameter - "a", "a,b,c" or "a/a1,b/b1,c/c1" */
+	errMsg := make([]string, 0, 1)
+	for _, s := range strings.Split(args, ",") {
+		reg := strings.Split(s, "/")
+		if len(reg[0]) == 0 {
+			continue
+		}
+
+		if _, ok := clsScopedRes[reg[0]]; ok {
+			/* A cluster scoped resource type is specified */
+			errMsg = append(errMsg, reg[0])
+		}
+	}
+
+	length := len(errMsg)
+
+	if length > 0 {
+		prefix := "--namespace (-n)"
+		if !nsSpecified {
+			prefix = "--all-namespaces"
+		}
+
+		postfix := ""
+		if length > 1 {
+			postfix = "s"
+		}
+		return fmt.Errorf("%s should not be specified for resource type%s: %s",
+					prefix, postfix, strings.Join(errMsg, " "))
+	} else {
+		return nil
+	}
+}
+
 func (o *DeleteOptions) Complete(f cmdutil.Factory, args []string, cmd *cobra.Command) error {
 	cmdNamespace, enforceNamespace, err := f.ToRawKubeConfigLoader().Namespace()
 	if err != nil {
 		return err
 	}
 
+	err = checkNamespaceWithArgs(f, args[0], cmd)
```

Comment:
Rejecting a command that specified -n along with cluster-scoped resources is not backwards compatible. at most, we could warn at this point. It would also be much simpler to check this after you have o.Result and o.Mapper resolved (or in the delete visit loop).

In Complete():
```
o.WarnClusterScope = enforceNamespace
```

in DeleteResult():
```
if !warnedClusterScope && o.WarnClusterScope && info.Mapping.Scope.Name() == meta.RESTScopeNameRoot {
	warnedClusterScope = true
	fmt.Fprintf(o.ErrOut, "Warning: deleting cluster-scoped resources, not scoped to the provided namespace\n")
}

response, err := o.deleteResource(info, options)
...
```

that would also give us a place to require an explicit opt-in in the future to delete cluster-scoped things in combination with an explicit --namespace arg (after a suitable warning/deprecation period)

Reply:
@liggitt 

     This approach is simple, but the warning message is appear alternately.

      Meanwhile, we don't need to show warning message for "default", "kube-system", "public" namespace who is rejected by server.


     BTW, do we need to show warning message if --all-namespaces is specified in the command line?

     I prefer to show one warning message for the command and handle --all-namespaces as --namespace.

./kubectl delete namespace -n test --all
Warning: deleting cluster-scoped resources <test>, not scoped to the provided namespace
namespace "test" deleted
Warning: deleting cluster-scoped resources <test1>, not scoped to the provided namespace
namespace "test1" deleted
Warning: deleting cluster-scoped resources <test2>, not scoped to the provided namespace
namespace "test2" deleted
Error from server (Forbidden): namespaces "default" is forbidden: this namespace may not be deleted
Error from server (Forbidden): namespaces "kube-public" is forbidden: this namespace may not be deleted
Error from server (Forbidden): namespaces "kube-system" is forbidden: this namespace may not be deleted

Reply:
The "warnedClusterScope = true" marker could let us show the warning once per invocation

Reply:
I could see skipping the warning if --all-namespaces was specified. Is that allowed in combination with an explicit --namespace arg?

Reply:
-n name and --all-namespaces={true|false} is allowed to specifed together in command.

Reply:
@liggitt @smarterclayton @yue9944882

   Diff has been updated and please help review it.